test/RaffleStore_test.js

```diff
@@ -93,18 +100,144 @@ describe("RaffleStore", function () {
 
-  // TODO: test entering a closed / pending_completion raffle
+  it("Should check the eth deposit matches the ticket price", async () => {
+    await testNft.connect(raffleOwner).approve(raffleStore.address, nftId)
+    await raffleStore.connect(raffleOwner).createRaffle(testNft.address, nftId, totalRaffleTickets, totalRafflePrice);
+    
+    await expect(
+      raffleStore.connect(rafflePlayer).enterRaffle(0, 1)
+    ).to.be.revertedWith("Ticket price not paid")
 
-  // TODO: test ticket purchase transfers eth
+    await expect(
+      raffleStore.connect(rafflePlayer).enterRaffle(0, 1, {
+        value: ticketPrice.div(2)
+      })
+    ).to.be.revertedWith("Ticket price not paid")
+    
+    await expect(
+      raffleStore.connect(rafflePlayer).enterRaffle(0, 1, {
+        value: ticketPrice
+      })
+    ).to.emit(raffleStore, "TicketsPurchased").withArgs(0, rafflePlayer.address, 1);
 
+    await expect(
+      raffleStore.connect(rafflePlayer).enterRaffle(0, 2, {
+        value: ticketPrice.mul(2)
+      })
+    ).to.emit(raffleStore, "TicketsPurchased").withArgs(0, rafflePlayer.address, 2);
+
+    await expect(
+      raffleStore.connect(rafflePlayer).enterRaffle(0, 2, {
+        value: ticketPrice
+      })
+    ).to.be.revertedWith("Ticket price not paid")
+  })
+
+  it("Should choose a winner when the last ticket is purchased", async () => {
+    await testNft.connect(raffleOwner).approve(raffleStore.address, nftId)
+    await raffleStore.connect(raffleOwner).createRaffle(testNft.address, nftId, totalRaffleTickets, totalRafflePrice);
+    
+    // TODO: check link was sent to vrfCoordinator
+    // const LinkToken = await deployments.get('LinkToken')
+    // linkToken = await ethers.getContractAt('LinkToken', LinkToken.address)
+
+    // await expect(
+    //   raffleStore.connect(rafflePlayer).enterRaffle(0, totalRaffleTickets, {
+    //     value: ticketPrice.mul(totalRaffleTickets)
+    //   })
+    // ).to.changeTokenBalances(
+    //   linkToken,
+    //   [raffleStore, vrfCoordinatorMock],
+    //   [-200, 200]
+    // );
+
+    await raffleStore.connect(rafflePlayer).enterRaffle(0, totalRaffleTickets, {
+      value: ticketPrice.mul(totalRaffleTickets)
+    })
+
+    await expect(
+      (await raffleStore.raffles(0)).status
+    ).to.equal(1)
+
+  })
+
+  it("Should not sell tickets while the raffle is pending completion", async () => {
+    await testNft.connect(raffleOwner).approve(raffleStore.address, nftId)
+    await raffleStore.connect(raffleOwner).createRaffle(testNft.address, nftId, totalRaffleTickets, totalRafflePrice);
+
+    await raffleStore.connect(rafflePlayer).enterRaffle(0, totalRaffleTickets, {
+      value: ticketPrice.mul(totalRaffleTickets)
+    })
+
+    await expect(
+      raffleStore.connect(rafflePlayer).enterRaffle(0, totalRaffleTickets, {
+        value: totalRafflePrice
+      })
+    ).to.be.revertedWith("Raffle no longer active")
+
+  })
+
+  it("Should award NFT to the winner", async () => {
```

Comment:
Nitpick: we're testing that the NFT is awarded to someone, not specifically the winner (there are no other players so our contract could be buggy and send the NFT to someone who isn't actually the winner and this test would pass).
```suggestion
  it("Should award the NFT when the final ticket is bought", async () => {
```

Reply:
This asserts that the token was transferred to the winning address (in this case the only user that played this raffle), so it would not pass if it was not transferred or transferred to a different address.  I still need to write tests for mapping randomness to winner and I wanted to test that logic separately

Reply:
Okay cool, ship ittt